NWNXLib/Services/Commands/Commands.cpp

```diff
@@ -0,0 +1,128 @@
+#include "Services/Commands/Commands.hpp"
+#include <stdio.h>
+#include <dlfcn.h>
+
+namespace NWNXLib {
+namespace Services {
+
+static Commands *g_commands;
+
+Commands::Commands()
+{
+    g_commands = this;
+    RegisterCommand("nwnx_commands_test", [](std::string& s){ LOG_NOTICE("Selftest command. Args: '%s'", s.c_str()); });
+}
+
+Commands::~Commands()
+{
+    g_commands = nullptr;
+
+}
+
+bool Commands::RegisterCommand(const std::string& cmd, CommandFunc func)
+{
+    auto it = m_commandMap.find(cmd);
+    if (it == m_commandMap.end())
+    {
+        m_commandMap[cmd] = func;
+        LOG_INFO("Registering command '%s'", cmd.c_str());
+        return true;
+    }
+    else
+    {
+        LOG_WARNING("Command '%s' already registered", cmd.c_str());
+        return false;
+    }
+}
+
+void Commands::UnregisterCommand(const std::string& cmd)
+{
+    LOG_INFO("Unregistering command '%s'", cmd.c_str());
+    m_commandMap.erase(cmd);
+}
+
+bool Commands::ScheduleCommand(const std::string& cmdline)
+{
+    std::string cmd, args;
+    size_t space = cmdline.find(' ');
+
+    if (space == std::string::npos) // No args
+    {
+        cmd = cmdline.substr(0, cmdline.length() - 1);
+        args = "";
+    }
+    else
+    {
+        cmd = cmdline.substr(0, space);
+        args = cmdline.substr(space+1, cmdline.length()-space-2);
+    }
+
+    auto it = m_commandMap.find(cmd);
+    if (it != m_commandMap.end())
+    {
+        std::lock_guard<std::mutex> guard(m_queueLock);
+        m_commandQueue.push_back(std::make_pair(cmd, args));
```

Comment:
Minor: You could `std::move` the strings in (and do the log above the command queue push) to avoid two string cpies. Maybe the optimiser will do this for you. You make be able to use `emplace_back` and remove the `std::make_pair` to avoid a third allocation but I don't know if that will work for sure.

Reply:
I'll do this in the cleanup pass. Thanks.